### PR TITLE
Ensure write error is included in exception message

### PR DESCRIPTION
**Pull request by jyemin**
Implement new CRUD specification prose test for write error details

JAVA-3993

### CODE REVIEW

File: driver-sync/src/test/functional/com/mongodb/client/CrudProseTest.java [hunk 1]
```diff
@@ -30,41 +35,83 @@
 import static java.lang.String.format;
 import static java.util.Arrays.asList;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
 import static org.junit.Assert.fail;
 import static org.junit.Assume.assumeTrue;
 
-// See https://github.com/mongodb/specifications/tree/master/source/change-streams/tests/README.rst#prose-tests
```
Comment (jyemin) [inline driver-sync/src/test/functional/com/mongodb/client/CrudProseTest.java:37]: This comment was copy pasta-d from elsewhere.

File: driver-sync/src/test/functional/com/mongodb/client/CrudProseTest.java [hunk 2]
```diff
@@ -30,41 +35,83 @@
...
-public class WriteConcernProseTest extends DatabaseTestCase {
+/**
+ * See https://github.com/mongodb/specifications/blob/master/source/crud/tests/README.rst#prose-tests
+ */
+public class CrudProseTest extends DatabaseTestCase {
```
Comment (jyemin) [inline driver-sync/src/test/functional/com/mongodb/client/CrudProseTest.java:47]: Renamed class to indicate the spec from which the prose tests came

File: driver-core/src/main/com/mongodb/MongoWriteException.java [hunk 3]
```diff
@@ -33,7 +33,7 @@
      * @param serverAddress the server address
      */
     public MongoWriteException(final WriteError error, final ServerAddress serverAddress) {
-        super(error.getCode(), error.getMessage(), serverAddress);
+        super("Write operation error on server " + serverAddress + ". Write error: " + error + ". ", serverAddress);
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/MongoWriteException.java:36]: Ensures that the whole write error, including details, will show up as part of the exception message.  This matches the behavior of MongoBulkWriteException.

**Comment by jyemin** [comment]
Converted to draft as tests are failing.

File: driver-sync/src/test/functional/com/mongodb/client/CrudProseTest.java [hunk 4]
```diff
@@ -30,41 +35,83 @@
...
             disableFailPoint();
         }
     }
 
+    /**
+     * 2. WriteError.details exposes writeErrors[].errInfo
+     */
+    @Test
+    public void testWriteErrorDetailsIsPropagated() {
+        assumeTrue(serverVersionAtLeast(3, 2));
```
Comment (stIncMale) [inline driver-sync/src/test/functional/com/mongodb/client/CrudProseTest.java:85]: JFYI: the spec [says _"Using a 5.0+ server"_](https://github.com/mongodb/specifications/commit/5141a8968a36cb4f29f203110c5d48aaaa281533#diff-0f732e30e0b0253a827e0fb457f5b27a1d80332db2e046d5a45587ba744e4455R257).

Comment (jyemin) [reply driver-sync/src/test/functional/com/mongodb/client/CrudProseTest.java:85]: Only one assertion (about details) is inappropriate for older servers, so just made that one conditional.

**Review by stIncMale — APPROVED** [review]

**Pull request merged**